### PR TITLE
segmentation violation when the function body lies near the end of a page

### PR DESCRIPTION
**Pull request by Streamlet**
I fixed it by calling Mprotect several times

### CODE REVIEW

File: replace_unix.go [hunk 1]
```diff
@@ -6,21 +6,24 @@ import (
 	"syscall"
 )
 
+func mprotectCrossPage(addr uintptr, len int, prot int) {
```
Comment (bouk) [inline replace_unix.go]: please change the `len` variable name to something else, as you're shadowing the builtin

File: replace_unix.go [hunk 2]
```diff
@@ -6,21 +6,24 @@ import (
...
+	pageSize := syscall.Getpagesize()
+	for p := pageStart(addr); p <= addr + uintptr(len); p += uintptr(pageSize) {
+		page := rawMemoryAccess(p, syscall.Getpagesize())
```
Comment (bouk) [inline replace_unix.go]: you can use `pageSize` here instead of calling `Getpagesize`

**Review by bouk — CHANGES_REQUESTED** [review]
Just some small notes

File: replace_unix.go (hunk 1 above)
Comment (Streamlet) [reply replace_unix.go]: ok, 3q for reviewing

**Comment by Streamlet** [comment]
I have fixed the above two issues

File: replace_unix.go (hunk 2 above)
Comment (Streamlet) [reply replace_unix.go]: ok, 3q

File: replace_unix.go [hunk 3]
```diff
@@ -6,21 +6,24 @@ import (
 	"syscall"
 )
 
+func mprotectCrossPage(addr uintptr, length int, prot int) {
+	pageSize := syscall.Getpagesize()
+	for p := pageStart(addr); p <= addr + uintptr(length); p += uintptr(pageSize) {
```
Comment (bouk) [inline replace_unix.go]: This should be < not <=

**Review by bouk — CHANGES_REQUESTED** [review]

File: replace_unix.go (hunk 3 above)
Comment (Streamlet) [reply replace_unix.go]: ah, yes, you are right.

**Comment by Streamlet** [comment]
fixed again

**Comment by wenjinhust** [comment]
bouk, your monkey cannot support go 1.12.4

**Pull request merged**